File: app/src/main/java/com/example/customview/WhoopCalendarView/CalendarView.java
package com.example.customview.WhoopCalendarView;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.customview.R;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;

public class CalendarView extends LinearLayout implements View.OnClickListener {
    LinearLayout header;
    ImageView btnPrev;
    ImageView btnNext;
    TextView txtDisplayDate;
    RecyclerView gridView;
    Calendar currentDate;
    int flag=0;
    Context context;

    CalendarUIAdapter calendarUIAdapter;

    int MIN_DISTANCE = 100;
    float downX = 0.0f;
    float downY = 0.0f;
    float upX = 0.0f;
    float upY = 0.0f;

    public CalendarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initControl(context, attrs);
        this.context=context;
    }

    private void assignUiElements() {
        // layout is inflated, assign local variables to components
        header = findViewById(R.id.calendar_header);
        btnPrev = findViewById(R.id.calendar_prev_button);
        btnNext = findViewById(R.id.calendar_next_button);
        txtDisplayDate = findViewById(R.id.date_display_date);
        gridView = findViewById(R.id.calendar_grid);
        btnPrev.setOnClickListener(this);
        btnNext.setOnClickListener(this);

    }

    /**
     * Load control xml layout
     */
    private void initControl(Context context, AttributeSet attrs) {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        inflater.inflate(R.layout.calendar_layout, this);
        assignUiElements();
        currentDate = Calendar.getInstance();
    }

    public void updateCalendar()
    {
        ArrayList<DateCell> cells = new ArrayList<>();
        Calendar calendar = (Calendar)currentDate.clone();
        calendar.set(Calendar.DAY_OF_MONTH, 1);
        calendar.setFirstDayOfWeek(Calendar.MONDAY);

        Calendar nextMonth = (Calendar) calendar.clone();
        nextMonth.add(Calendar.MONTH,1);

        while (calendar.get(Calendar.DAY_OF_WEEK) != Calendar.MONDAY) {
            calendar.add(Calendar.DATE, -1);
        }

        for(int i=0;i<38;i++){

            cells.add(new DateCell(calendar.getTime(),
                    "#000000",
                    R.color.transparent, i,
                    isCurrentMonth(calendar),
                    false,
                    false));

            calendar.add(Calendar.DAY_OF_MONTH, 1);

            if (calendar.compareTo(nextMonth) >= 0) {
                break;
            }
        }

    calendarUIAdapter = new CalendarUIAdapter(cells);
    calendarUIAdapter.setSelectedInstance(currentDate);
    calendarUIAdapter.setFlag(flag);
    gridView.setLayoutManager(new GridLayoutManager(context, 7, RecyclerView.VERTICAL, false));
    gridView.setAdapter(calendarUIAdapter);

        gridView.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN: {
                        downX = event.getX();
                        downY = event.getY();
                        return true;
                    }
                    case MotionEvent.ACTION_UP: {
                        upX = event.getX();
                        upY = event.getY();

                        float deltaX = downX - upX;
                        float deltaY = downY - upY;

                        if (Math.abs(deltaX) > MIN_DISTANCE) {
                            // left or right
                            if (deltaX < 0) {
                                onLeftToRightSwipe(v);
                                return true;
                            }
                            if (deltaX > 0) {
                                onRightToLeftSwipe(v);
                                return true;
                            }
                        }

                        if (Math.abs(deltaY) > MIN_DISTANCE) {
                            // top or down
                            if (deltaY < 0) {
                                onTopToBottomSwipe(v);
                                return true;
                            }
                            if (deltaY > 0) {
                                onBottomToTopSwipe(v);
                                return true;
                            }
                        }

                        return false; // no swipe horizontally and no swipe vertically
                    }
                }
                return false;
            }
        });


    SimpleDateFormat sdf = new SimpleDateFormat("EEEE,MMMM yyyy");
    String[] dateToday = sdf.format(currentDate.getTime()).split(",");
    txtDisplayDate.setText(dateToday[1]);
    }


    @Override
    public void onClick(View v) {
        if(v.getId() == R.id.calendar_prev_button){

            updateTheCalendar(-1);

        } else {
            updateTheCalendar(1);

        }
        calendarUIAdapter.notifyDataSetChanged();
    }

    public void updateTheCalendar(int month){
        currentDate.add(Calendar.MONTH,month);
        updateCalendar();
    }

    public boolean isCurrentMonth(Calendar calendar){

        int month = calendar.get(Calendar.MONTH);
        int year = calendar.get(Calendar.YEAR);

        if (month == currentDate.get(Calendar.MONTH) && year == currentDate.get(Calendar.YEAR)) {
            return true;
        } else {
            return false;
        }
    }


    public void setFlag(int flag) {
        this.flag = flag;
        calendarUIAdapter.setFlag(flag);
        calendarUIAdapter.notifyDataSetChanged();
    }

    public void onRightToLeftSwipe(View v) {
        updateTheCalendar(1);
        Toast.makeText(v.getContext(), "RightToLeftSwipe", Toast.LENGTH_SHORT).show();
        // activity.doSomething();
    }

    public void onLeftToRightSwipe(View v) {
        updateTheCalendar(-1);
        Toast.makeText(v.getContext(), "LeftToRightSwipe", Toast.LENGTH_SHORT).show();
        // activity.doSomething();
    }

    public void onTopToBottomSwipe(View v) {

        Toast.makeText(v.getContext(), "onTopToBottomSwipe", Toast.LENGTH_SHORT).show();
        // activity.doSomething();
    }

    public void onBottomToTopSwipe(View v) {
        Toast.makeText(v.getContext(), "onBottomToTopSwipe", Toast.LENGTH_SHORT).show();
        // activity.doSomething();
    }
}

File: app/src/main/java/com/example/customview/WhoopCalendarView/CalendarView.kt
package com.example.customview.WhoopCalendarView

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.customview.R
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.abs

class CalendarView(context: Context, attrs: AttributeSet) :
    LinearLayout(context, attrs), View.OnClickListener {

    private lateinit var header: LinearLayout
    private lateinit var btnPrev: ImageView
    private lateinit var btnNext: ImageView
    private lateinit var txtDisplayDate: TextView
    private lateinit var gridView: RecyclerView
    private lateinit var calendarUIAdapter: CalendarUIAdapter
    private lateinit var currentDate: Calendar
    var flag : Int = 0

    private var MIN_DISTANCE = 100
    private var downX = 0.0f
    private var downY = 0.0f
    private var upX = 0.0f
    private var upY = 0.0f

    private fun assignUiElements() {
        header = findViewById(R.id.calendar_header)
        btnPrev = findViewById(R.id.calendar_prev_button)
        btnNext = findViewById(R.id.calendar_next_button)
        txtDisplayDate = findViewById(R.id.date_display_date)
        gridView = findViewById(R.id.calendar_grid)
        btnPrev.setOnClickListener(this)
        btnNext.setOnClickListener(this)
    }

    /**
     * Load control xml layout
     */

    init {
        initControl(context)
    }
    private fun initControl(
        context: Context
    ) {

        val inflater =
            context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.calendar_layout, this)
        assignUiElements()
        currentDate = Calendar.getInstance()
    }

    fun updateCalendar() {
        val cells = ArrayList<DateCell>()
        val calendar = currentDate.clone() as Calendar
        calendar[Calendar.DAY_OF_MONTH] = 1
        calendar.firstDayOfWeek = Calendar.MONDAY
        val nextMonth = calendar.clone() as Calendar
        nextMonth.add(Calendar.MONTH, 1)
        while (calendar[Calendar.DAY_OF_WEEK] != Calendar.MONDAY) {
            calendar.add(Calendar.DATE, -1)
        }
        for (i in 0..37) {
            cells.add(
                DateCell(
                    calendar.time,
                    "#000000",
                    R.color.transparent,i,
                    isCurrentMonth(calendar)
                )
            )
            calendar.add(Calendar.DAY_OF_MONTH, 1)
            if (calendar >= nextMonth) {
                break
            }
        }
        while (calendar[Calendar.DAY_OF_WEEK] != Calendar.MONDAY) {
            cells.add(
                DateCell(
                    calendar.time,
                    "#000000",
                    R.color.transparent, cells.size + 1,
                    isCurrentMonth(calendar)
                )
            )
            calendar.add(Calendar.DATE, 1)
        }
        calendarUIAdapter = CalendarUIAdapter(cells)
        calendarUIAdapter.selectedInstance = currentDate
        calendarUIAdapter.flag = flag
        gridView.layoutManager = GridLayoutManager(context, 7, RecyclerView.VERTICAL, false)
        gridView.adapter = calendarUIAdapter
        gridView.setOnTouchListener(OnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    downX = event.x
                    downY = event.y
                    return@OnTouchListener true
                }
                MotionEvent.ACTION_UP -> {
                    upX = event.x
                    upY = event.y
                    val deltaX = downX - upX
                    val deltaY = downY - upY
                    if (abs(deltaX) > MIN_DISTANCE) { // left or right
                        if (deltaX < 0) {
                            onLeftToRightSwipe()
                            return@OnTouchListener true
                        }
                        if (deltaX > 0) {
                            onRightToLeftSwipe()
                            return@OnTouchListener true
                        }
                    }
                    if (abs(deltaY) > MIN_DISTANCE) { // top or down
                        if (deltaY < 0) {
                            onTopToBottomSwipe()
                            return@OnTouchListener true
                        }
                        if (deltaY > 0) {
                            onBottomToTopSwipe()
                            return@OnTouchListener true
                        }
                    }
                    return@OnTouchListener false // no swipe horizontally and no swipe vertically
                }
            }
            false
        })
        val sdf = SimpleDateFormat("EEEE,MMMM yyyy")
        val dateToday =
            sdf.format(currentDate.time).split(",").toTypedArray()
        txtDisplayDate.text = dateToday[1]
    }

    override fun onClick(v: View) {
        if (v.id == R.id.calendar_prev_button) {
            updateTheCalendar(-1)
        } else {
            updateTheCalendar(1)
        }
        calendarUIAdapter.notifyDataSetChanged()
    }

    private fun updateTheCalendar(month: Int) {
        currentDate.add(Calendar.MONTH, month)
        updateCalendar()
    }

    private fun isCurrentMonth(calendar: Calendar): Boolean {
        val month = calendar[Calendar.MONTH]
        val year = calendar[Calendar.YEAR]
        return month == currentDate[Calendar.MONTH] && year == currentDate[Calendar.YEAR]
    }



    fun setType( f:Int){
        this.flag = f
        calendarUIAdapter.flag = f
        calendarUIAdapter.notifyDataSetChanged()
    }


    private fun onRightToLeftSwipe() {
        updateTheCalendar(1)
    }

    private fun onLeftToRightSwipe() {
        updateTheCalendar(-1)
    }

    private fun onTopToBottomSwipe() {
        updateTheCalendar(-1)
    }

    private fun onBottomToTopSwipe() {
        updateTheCalendar(1)
    }

}
